django_advanced_queries/covid_19/models.py

```diff
@@ -2,12 +2,285 @@
+        ).filter(
+            is_dead=MedicalExaminationResult.RESULT_DEAD,
+            reason_of_dead=MedicalExaminationResult.RESULT_CORONA
+        )
+
+
+class DepartmentManager(models.Manager):
+    def annotate_avg_age_of_patients(self):
+        return self.annotate(
+            avg_age_of_patients=models.Avg("patients_details__person__age")
+        )
+
+
+class HospitalManager(models.Manager):
+    def annotate_by_num_of_hospital_workers_in_risk_of_corona(self):
+        # Someone who is in risk group of corona is person that is older
+        # than 60
+        workers_in_risk = HospitalWorker.objects.filter(
+            person__age__gte=60,
+            department__hospital=models.OuterRef("id")
+        ).values("department__hospital", "id").distinct()
+
+        # Count with risky counter.
+        risk_per_hospital = workers_in_risk.annotate(
+            risky_count=models.Count("department__hospital")
+        ).values("risky_count")
+
+        # Perform group by aggregation with workers in risk count.
+        hospitals_with_risk_counter = self.annotate(
+            num_of_hospital_workers_in_risk_of_corona=
+            models.Subquery(risk_per_hospital,
+                            output_field=models.IntegerField())
+        )
+
+        return hospitals_with_risk_counter
+
+    def annotate_by_num_of_dead_from_corona(self):
+        # Filter all corona dead persons.
+        dead_corona_patient_details = Patient.objects.get_dead_details()
+
+        # Annotate with dead corona patients counter.
+        hospitals_per_dead_patient = dead_corona_patient_details.filter(
+            department__hospital=models.OuterRef("id"),
+        ).values(
+            "department__hospital"
+        ).annotate(
+            count=models.Count("id")
+        ).values("count")
+
+        hospital_with_corona_dead_details = self.annotate(
+            # Coalesce to prevent None values (Take the first non-none value).
+            num_of_dead_from_corona=Coalesce(
+                models.Subquery(hospitals_per_dead_patient,
+                                output_field=
+                                models.IntegerField()), 0)
+        )
+
+        return hospital_with_corona_dead_details
+
+    def annotate_hospitals_with_time_of_first_corona_sick(self):
+        # Filter by parent query id and corona result -> Get the earliest.
+        first_corona_time = MedicalExaminationResult.objects.filter(
+            patient__department__hospital=models.OuterRef("id"),
+            result=MedicalExaminationResult.RESULT_CORONA,
+        ).order_by("time").values("time")[:1]
+
+        hospital_with_corona_dead_details = self.annotate(
+            first_corona_time=models.Subquery(first_corona_time)
+        )
+
+        return hospital_with_corona_dead_details
+
+
+class HospitalWorkerManager(models.Manager, SickPersonsMixin):
+    """Custom hospital worker model Queryset manager."""
+
+    def get_queryset(self):
```

Comment:
overriding the default behaviour is not recommended since it cause fetching person details everytime everywhere.
In this case use select_related in the test  itself (I know its not clear from the test itself)

Reply:
Done #1